### PR TITLE
[WIP] lxd: add package

### PR DESCRIPTION
**Pull request by aparcar**
Maintainer: @feckert 

Description:

currently doesn't compile, uploaded here to seek advise from other Go maintainers.

Authored by: Florian Eckert <fe@dev.tdt.de>
Signed-off-by: Paul Spooren <mail@aparcar.org>

### CODE REVIEW

**Comment by karlp** [comment]
So what's WIP abotu it?  what sort of feedback ar eyou looking for at this stage?  this is all stuff you could put in the PR description...

File: utils/lxd/Makefile [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+PKG_BUILD_DEPENDS:=golang/host
+PKG_BUILD_PARALLEL:=1
+
+GO_PKG:=github.com/lxc/lxd
+GO_PKG_BUILD_PKG:=github.com/lxc/lxd/lxd/
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/../feeds/packages/lang/golang/golang-package.mk
+
+TAR_CMD:=$(HOST_TAR) -C $(PKG_BUILD_DIR) --strip-components 1 $(TAR_OPTIONS)
```
Comment (karlp) [inline utils/lxd/Makefile]: serious?  that's an unusual approach. normally people just edit hte PKG_SOURCE directories appropriately.  I knwo it's "equivalent" but this is much more intrusively reaching into the build guts.

File: utils/lxd/Makefile [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	   +acl \
+	   +lxc \
+	   +liblxc \
+	   +squashfs-tools-unsquashfs \
+	   +xz \
+	   +xz-utils \
+	   +shadow-newuidmap \
+	   +shadow-newgidmap \
+	   +tar \
+	   +rsync
```
Comment (karlp) [inline utils/lxd/Makefile]: If you'r egoing to use +=, why not avoid the \ mess and just add each one?  Avoids any merge problems with contexts and '\' chars.

**Comment by aparcar** [comment]
Hi @karlp , thanks for you kind words of advise. @feckert send me a Makefile via email asking for help, so I though to keep the process on GitHub instead of using diffs around.

File: utils/lxd/Makefile (hunk 2 above)
Comment (aparcar) [reply utils/lxd/Makefile]: Do you mean to use `DEPENDS+=` for each single dependency instead of \?

Comment (karlp) [reply utils/lxd/Makefile]: yes, it avoids the context mangling that happens with some lines having \ and others not, and you're using += on DEPENDS anyway.  I don't _really_ mind, I've just seen enough merge errors in my life form having to modify extra lines to fix up the end of line escaping, and from people trying to comment out a dependency and failing, and += just magically makes all lines self contained.

**Comment by feckert** [comment]
@aparcar I made some change to the makefile.
[changes.txt](https://github.com/openwrt/packages/files/2992561/changes.txt)
please have a look

I still get some erros but it is starting compiling YEAH

>github.com/CanonicalLtd/go-dqlite/internal/bindings
src/github.com/CanonicalLtd/go-dqlite/internal/bindings/cluster.go:7:10: fatal error: dqlite.h: No such file or directory
 #include <dqlite.h>
          ^~~~~~~~~~
compilation terminated.

I don't know if this is the right way either.


For now it is only a try and error development!

**Comment by aparcar** [comment]
@feckert I feel like redefining the functions is wrong, maybe @jefferyto can help us out. I see that `go get <url>` is required but somehow not run by golang-package.mk, but can't tell what we're doing wrong...

**Comment by jefferyto** [comment]
The reason why `go get` here "works" is because it downloads all of the package's Go dependencies. The OpenWrt Go build system explicitly does not do that, because the dependencies are not versioned.

The OpenWrt Go build system was based on Debian's dh_golang, and like dh_golang, expects each Go dependency to be packaged separately (and thus pinned to a version or commit). When the Makefile includes a `DEPENDS` on another Go package, the build system "pulls in" the dependency's source code and makes it available for use. An example of a package using this system is obfs4proxy (9c6d03db9eca778eaa27026a9d825d66fa20a101). (obfs4proxy has since been ported to a Go module and so now the Go compiler manages the dependencies.)

If the upstream package was a Go module or used a `vendor` directory, then there would be no problem, but as it stands now, unless you or someone else packages the dozens of Go dependencies (if you do the `go get` command, you will see all the packages it downloads), I don't see how this can be compiled under the existing build system.

(Similarly, the `make deps` command actually builds two C libraries, also unversioned. These would need to be packaged as OpenWrt packages as well.)

(And on a related note, please don't blindly copy upstream build instructions into a OpenWrt makefile and expect it to work. I suggest opening the upstream build script / makefile / etc., reading and understanding what it is they are doing, and then adapting/translating the code. In fact, I suggest reading golang-package.mk and the other helper Go makefiles and understanding what they are doing.)

**Comment by jefferyto** [comment]
If you really, really want to build this, it looks like the tarball bundles all(?) of its Go dependencies in the `dist/src` directory. You can do something like the ["ugly workaround"](https://github.com/lxc/lxd-pkg-ubuntu/blob/8dc4a41e780000c323e119ba14f4463b4c222c8e/debian/rules#L24) that the (old) Ubuntu package does and copy the dependencies into the right folder.

**Comment by feckert** [comment]
>The reason why go get here "works" is because it downloads all of the package's Go dependencies. The OpenWrt Go build system explicitly does not do that, because the dependencies are not versioned.

May it be possible to ping the lxd maintainer to add there dependencies as a go.mod file?

>The OpenWrt Go build system was based on Debian's dh_golang, and like dh_golang ...

I already notice this thanks, but it is a hard work to add all this dependencies!!

> If the upstream package was a Go module or used a vendor directory, then there would be no ...

If we package all the dependencies then we have the same problem as in python!
Different binaries in go need perhaps different version from the depended libraries.
So if we package all of them we perhaps end in a dependencies hell as in python because diffrent go bin files need different lib versions.

>(Similarly, the make deps command actually builds two C libraries, also unversioned. These would need to be packaged as OpenWrt packages as well.)

That's what I thought, too. That would have been my next step.

> (And on a related note, please don't blindly copy upstream build instructions into a OpenWrt ...

These are my first attempts with golang. I have already looked at the other packages. I thought it's no problem to package lxd for openwrt but I was wrong. There is nothing else left for me to get deeper into the topic.

> If you really, really want to build this, it looks like the tarball bundles all(?) of its Go dependencies in the dist/src directory. ...

I guess we'll have no choice to try this way first.

Thanks for help

**Comment by jefferyto** [comment]
> May it be possible to ping the lxd maintainer to add there dependencies as a go.mod file?

I think it's worth trying to contribute this to upstream.

**Comment by feckert** [comment]
After playing around with lxd in OpenWrt to get it to run, here are my findings:

Lxd needs two libraries to build
sqlite https://github.com/CanonicalLtd/sqlite
dqlite https://github.com/CanonicalLtd/dqlite 

sqlite:
Is already available in OpenWrt but can **NOT** be used because lxd requires a patched version.

dqlite:
The library dqlite  compiles on OpenWrt with the following Makefile
[Makefile.txt](https://github.com/openwrt/packages/files/3028744/Makefile.txt)

Also building the lxd go dependencies as described
> If you really, really want to build this, it looks like the tarball bundles all(?) of its Go dependencies in the `dist/src` directory. You can do something like the ["ugly workaround"](https://github.com/lxc/lxd-pkg-ubuntu/blob/8dc4a41e780000c323e119ba14f4463b4c222c8e/debian/rules#L24) that the (old) Ubuntu package does and copy the dependencies into the right folder.

does work with the following change in the Build/Configure callback in den lxd Makefile

> define Build/Configure
> 	$(call GoPackage/Build/Configure)
> 
> 	$(CP) $(GO_PKG_BUILD_DIR)/src/* $(GO_PKG_TMP_DIR)
> 	$(RM) -rf $(GO_PKG_BUILD_DIR)/src
> 	$(CP) $(PKG_BUILD_DIR)/dist/src $(GO_PKG_BUILD_DIR)
> 	$(RM) -rf $(GO_PKG_BUILD_DIR)/src/github.com/lxc
> 	$(CP) $(GO_PKG_TMP_DIR)/github.com/* $(GO_PKG_BUILD_DIR)/src/github.com
> 	$(RM) -rf $(GO_PKG_TMP_DIR)/github.com
> endef

Since lxd is based on a patched sqlite I still have problems compiling lxd for OpenWrt.
I don't know how to solve this in Openwrt.
**Any help is welcome.**

**Comment by jefferyto** [comment]
Not sure if you have seen CanonicalLtd/dqlite#91, but since the patch only adds functionality, perhaps it can be added to the existing sqlite package, or a separate package for the forked library can be added that `PROVIDES` libsqlite3.

**Comment by feckert** [comment]
For the test I removed the original sqlite3 and replaced this with the patched sqlite from canonical. 
https://github.com/CanonicalLtd/sqlite
The patched sqlite is needed for the lxd service.
With this pullrequest change I was able to compile the lxc client binary successfully.

But I have still problems to compile the lxd service. I always get the following error.

> /home/feckert/workspace/lede/LDM-master-64/build/openwrt/staging_dir/toolchain-x86_64_gcc-7.4.0_musl/lib/gcc/x86_64-openwrt-linux-musl/7.4.0/../../../../x86_64-openwrt-linux-musl/bin/ld: /home/feckert/workspace/lede/LDM-master-64/build/openwrt/staging_dir/target-x86_64_musl/usr/lib/libdqlite.so: undefined reference to `dqlite__flip64'
> /home/feckert/workspace/lede/LDM-master-64/build/openwrt/staging_dir/toolchain-x86_64_gcc-7.4.0_musl/lib/gcc/x86_64-openwrt-linux-musl/7.4.0/../../../../x86_64-openwrt-linux-musl/bin/ld: /home/feckert/workspace/lede/LDM-master-64/build/openwrt/staging_dir/target-x86_64_musl/usr/lib/libdqlite.so: undefined reference to `dqlite__flip32'

I don know what I am doing wrong. I have also talked to @freeekanayaka that we have to set the tags **-tags libsqlite3** on **go install**. For the test I have adding this to the following lines.https://github.com/openwrt/packages/blob/990b083990374e717f0d7fc993e01f0648462d67/lang/golang/golang-package.mk#L283-L288

This does not solve my problem either. So it looks like I'm using a version of dqlite that doesn't have this function. I have found a change on the dqlite repo which Looks suspicious.
https://github.com/CanonicalLtd/dqlite/commit/bfa4a14a7ed1ad1cd74c27431758a018722c0848
But I think I use the correct version from the manifest file extracted from the lxd sources.

**Comment by freeekanayaka** [comment]
Hi Florian,

Florian Eckert <notifications@github.com> writes:

[...]
> I don know what I am doing wrong. I have also talked to @freeekanayaka that we have to set the tags **-tags libsqlite3** on **go install**. For the test I have adding this to the following lines.https://github.com/openwrt/packages/blob/990b083990374e717f0d7fc993e01f0648462d67/lang/golang/golang-package.mk#L283-L288
>
> This does not solve my problem either. So it looks like I'm using a version of dqlite that doesn't have this function. I have found a change on the dqlite repo which Looks suspicious.
> https://github.com/CanonicalLtd/dqlite/commit/bfa4a14a7ed1ad1cd74c27431758a018722c0848
> But I think I use the correct version from the manifest file extracted
> from the lxd sources.

That change is unrelated to your problem. It might be a linker/compiler
issue. What version f gcc and libc are you using exactly?

Free

**Comment by feckert** [comment]
@freeekanayaka thanks for feedback
I am using the OpenWrt master which uses the following versions on the x86_64 target.

- **gcc** is version 7.40

- libc is the **musl** with version 1.1.21

**Comment by diizzyy** [comment]
What's the status?

**Comment by feckert** [comment]
@diizzyy I think we could close this because I am using now the docker-ce package.
There were too many stumbling blocks to get the package up and running, especially with sqlite database. I talked of topic with the developers of lxd and their patches are now not upstreamable for sqlite. So I moved on to docker-cd

**Comment by aparcar** [comment]
Fine for me.

**Pull request closed without merging**